File: src/lib/validation.ts
/**
 * Client-side validation utilities
 *
 * These functions use Tauri commands that delegate to k8s-gui-common/src/validation.rs,
 * ensuring consistency across all projects (src-tauri, auth-server, and frontend).
 *
 * All validation logic is centralized in k8s-gui-common for a single source of truth.
 */

import * as commands from "@/generated/commands";
import { normalizeTauriError } from "./error-utils";

export interface ValidationResult {
  isValid: boolean;
  error?: string;
}

/**
 * Validate email format
 *
 * Uses Tauri command that delegates to k8s-gui-common/src/validation.rs::validate_email
 * for consistency across all projects.
 *
 * @param email - Email address to validate
 * @returns Validation result with isValid flag and optional error message
 */
export async function validateEmail(email: string): Promise<ValidationResult> {
  try {
    await commands.validateEmailCommand(email);
    return { isValid: true };
  } catch (error) {
    return {
      isValid: false,
      error: normalizeTauriError(error),
    };
  }
}

/**
 * Password strength requirements
 *
 * Synchronized with: k8s-gui-common/src/validation.rs::PasswordRequirements
 * These values must match the Rust implementation.
 */
export const PASSWORD_REQUIREMENTS = {
  minLength: 8,
  maxLength: 128,
  requireLowercase: true,
  requireUppercase: true,
  requireDigit: true,
  requireSpecial: true,
  specialChars: "!@#$%^&*()_+-=[]{}|;:,.<>?",
};

/**
 * Validate password strength
 *
 * Uses Tauri command that delegates to k8s-gui-common/src/validation.rs::validate_password
 * for consistency across all projects.
 *
 * @param password - Password to validate
 * @returns Validation result with isValid flag and optional error message
 */
export async function validatePassword(password: string): Promise<ValidationResult> {
  try {
    await commands.validatePasswordCommand(password);
    return { isValid: true };
  } catch (error) {
    return {
      isValid: false,
      error: normalizeTauriError(error),
    };
  }
}

/**
 * Validate license key format (UUID format)
 *
 * Uses Tauri command that delegates to k8s-gui-common/src/validation.rs::validate_license_key
 * for consistency across all projects.
 *
 * @param licenseKey - License key to validate (must be in UUID format)
 * @returns Validation result with isValid flag and optional error message
 */
export async function validateLicenseKey(licenseKey: string): Promise<ValidationResult> {
  try {
    await commands.validateLicenseKeyCommand(licenseKey);
    return { isValid: true };
  } catch (error) {
    return {
      isValid: false,
      error: normalizeTauriError(error),
    };
  }
}

File: src-tauri/src/commands/validation.rs
//! Validation commands
//!
//! Provides validation functions that can be called from the frontend.
//! All validation logic is centralized in k8s-gui-common to ensure consistency.

use crate::error::{Error, Result};
use k8s_gui_common::validation::{validate_email, validate_license_key, validate_password};

/// Validate email format
///
/// Returns Ok(()) if email is valid, Err(String) with error message otherwise.
#[tauri::command]
pub fn validate_email_command(email: String) -> Result<()> {
    validate_email(&email).map_err(|e| Error::InvalidInput(e))
}

/// Validate password strength
///
/// Returns Ok(()) if password meets requirements, Err(String) with error message otherwise.
#[tauri::command]
pub fn validate_password_command(password: String) -> Result<()> {
    validate_password(&password).map_err(|e| Error::InvalidInput(e))
}

/// Validate license key format (UUID)
///
/// Returns Ok(()) if license key is in valid UUID format, Err(String) with error message otherwise.
#[tauri::command]
pub fn validate_license_key_command(license_key: String) -> Result<()> {
    validate_license_key(&license_key).map_err(|e| Error::InvalidInput(e))
}
